### PR TITLE
Fix test_distributions when WITH_SCALARS.

### PR DESCRIPTION
**Pull request by gchanan**

### CODE REVIEW

File: torch/distributions/categorical.py [hunk 1]
```diff
@@ -73,11 +73,19 @@ def param_shape(self):
 
     @property
     def mean(self):
-        return self.probs.new([float('nan')]).expand(self._extended_shape())
+        # verbose because of differening Variable/Tensor apis and lack of dtypes
+        if isinstance(self.probs, Variable):
+            return variable(float('nan')).type_as(self.probs).expand(self._extended_shape())
```
Comment (fritzo) [inline torch/distributions/categorical.py]: Does `.type_as()` place its result on the correct cuda device?

File: torch/distributions/pareto.py [hunk 2]
```diff
@@ -29,7 +29,7 @@ class Pareto(TransformedDistribution):
 
     def __init__(self, scale, alpha):
         self.scale, self.alpha = broadcast_all(scale, alpha)
-        base_dist = Exponential(alpha)
+        base_dist = Exponential(self.alpha)
```
Comment (fritzo) [inline torch/distributions/pareto.py]: cc @alicanb

File: torch/distributions/categorical.py (hunk 1 above)
Comment (gchanan) [reply torch/distributions/categorical.py]: apparently not.

**Comment by gchanan** [comment]
ok this works now both with and without scalars; I'll think of a workaround for the type_as issue.

File: torch/distributions/categorical.py (hunk 1 above)
Comment (fritzo) [reply torch/distributions/categorical.py]: I always use `probs.new([float('nan')]).expand(...)`. What is the new syntax we can use for scalars to get correct device placement?

File: torch/distributions/pareto.py [hunk 3]
```diff
@@ -29,8 +29,8 @@ class Pareto(TransformedDistribution):
 
     def __init__(self, scale, alpha):
         self.scale, self.alpha = broadcast_all(scale, alpha)
-        base_dist = Exponential(alpha)
-        transforms = [ExpTransform(), AffineTransform(loc=0, scale=scale)]
+        base_dist = Exponential(self.alpha)
+        transforms = [ExpTransform(), AffineTransform(loc=0, scale=self.scale)]
```
Comment (fritzo) [inline torch/distributions/pareto.py]: @alicanb pointed out that if `scale` and `alpha` are both scalars, then the 0 here will broadcast everything up to dimension-1. Does it seem reasonable to use `loc=variable(0.0)` here instead, at least until Python numbers become 0-dimensional?

Comment (alicanb) [reply torch/distributions/pareto.py]: `self.scale` is already the output of `broadcast_all` so I think it's safe now

Comment (fritzo) [reply torch/distributions/pareto.py]: but until `WITH_SCALARS=1`, 0 may be a higher-dimensional tensor than `self.scale` and therefore this broadcasting is wrong

Comment (gchanan) [reply torch/distributions/pareto.py]: Won't self.scale be at least 1-d?

Comment (fritzo) [reply torch/distributions/pareto.py]: Oh maybe I misunderstood. Is the following true?
- If `WITH_SCALARS=0` then tensors have at least 1 dim and python numbers have 1 dim
- If `WITH_SCALARS=1` then tensors may have 0 dims, and python numbers have 0 dims

If so then @alicanb is right and this is fine.

Comment (gchanan) [reply torch/distributions/pareto.py]: That's right if by "python numbers have X dim" you mean python numbers get broadcasted to a tensor of X dims in broadcast_all; python numbers themselves don't have dims.

**Comment by fritzo** [comment]
Here's a fix for `OneHotCategorical`:
```diff
diff --git a/torch/distributions/relaxed_categorical.py b/torch/distributions/relaxed_categorical.py
index 1961aeb0..54538f1d 100644
--- a/torch/distributions/relaxed_categorical.py
+++ b/torch/distributions/relaxed_categorical.py
@@ -65,7 +65,7 @@ class ExpRelaxedCategorical(Distribution):
     K = self._categorical._num_events
     self._validate_log_prob_arg(value)
     logits, value = broadcast_all(self.logits, value)
-    log_scale = (self.logits.new(1).fill_(K).lgamma() -
+    log_scale = (self.temperature.new(self.temperature.shape).fill_(K).lgamma() -
           self.temperature.log().mul(-(K - 1)))
     score = logits - value.mul(self.temperature)
     score = (score - log_sum_exp(score)).sum(-1)
```

**Comment by gchanan** [comment]
This now includes https://github.com/pytorch/pytorch/pull/5144.

**Review by colesbury — APPROVED** [review]
